Moving decision functionality into client (part 1)

datanoh opened the pull request:
### Summary of Changes
Significant refactor of AwaitingDecision server class and related client code. See the gamestate-example.hjson file for an overview of what decision data will look like in the client. Added new files to the client 'actionSelectionDecisions.js' and 'decisions.js'.

Moved several key components of decision structure into the client:
- The decision of whether or not to use the "cardSelectionDialog" element for a card selection is now determined by the client (decisions.js, getUseDialog method).
- The use of "temp" names to identify cards in the card selection dialog has been moved to the client (see decisions.js, CardSelectionDecision class, getDivIdFromCardId and getCardIdFromDivId methods). The server will send actual cardIds and expects actual cardIds back. The "temp" names also now include the actual cardId within them, so for example cardId 27 becomes "temp27" instead of "temp0". I'm also very intentionally not calling these "temp" values cardIds anywhere in the decisions.js code. (Strictly speaking I don't think they're the id properties of the div elements either, but meh)
- The server no longer sends blueprintIds and imageUrls to the client as part of card selection decisions. It just sends the cardIds, and the client can get the other elements from the serialized game state.
- The server is no longer sending card data to the client at all for an action selection decision. It's just sending a list of actions, in the same format being used in actionResults.js. (See actionSelectionDecisions.js, getActionInitiationCardActionMap function) This same function is assigning the texts to be used in the action choice context menu entirely within the client, except for the USE_GAME_TEXT action type. This action type is reserved for one-off actions that don't have their own specific class in the server.

Other changes:
- Removed the decision branch that selected actions from the card selection dialog.
- Action types ADD_CARDS_TO_PRESEED_STACK and REMOVE_CARDS_FROM_PRESEED_STACK were refactored both in the server and their client implementation in actionResults.js.
- Added NULLIFY action type to actionResults.js (this action type was added in a recent PR)
- Removed the decision branch for action selection that handled actions performed by cards not in play. Pretty sure I accidentally broke this a long time ago, and we shouldn't need it for a while.

### Testing Performed
Still passing all automated tests. I loaded up a game a couple times to try different selections to make sure they still worked, including seeding missions and dilemmas, the AMS download, normal card plays, and beaming.

### Benefits
Overall this is aligned with the server/client relationship we're aiming for. It also helps pave the way to introduce "compound decisions". At some point I'd like to start sending the client "action initiation" decisions. The best example of this is beaming. Instead of a series of back-and-forth action selection and card selection decisions, the initial action selection could just include all the data necessary for the client to decide how to create the "beam" action.

### Risks
- Some card selections that didn't previously use the card selection dialog are using it now. (For example, selecting personnel to beam and selecting dilemmas to seed.) This can be adjusted as desired in decisions.js.
- A couple of the user messages are less specific for now. Where observed, these were noted with a "TODO" comment in actionSelectionDecisions.js. For example:
  - The user prompt for selecting a mission to seed dilemmas under used to say "Select one of your missions", "Select one of your opponent's missions", etc. It now has a more general message that doesn't specify which missions are selectable. This could be added back in using elements already available in the serialized game state.
  - The user prompt in the action choice context menu for Attention All Hands used to say "report a personnel for free" and now just uses a generic "play card". This more specific language can be restored by adding a few elements to the action data in the serialized game state.
- The limited server data for card selection data may cause issues if selecting cards with limited player visibility. This is already an issue for other features, and I have some thoughts on how to address it overall in the game state data.
- Some risks already existed before this PR that weren't addressed. For example, I don't think there are any checks in the client that a card selection decision doesn't have multiple copies of the same card, or that the validCombinations has a list of compatible cards for every card in the selection.

### Possible Improvements to this PR
- The "temp" card div names could be removed entirely within the client. I _think_ this would just need the following, but I haven't tested this. Not a high priority.
  - updating the getCardDivFromId function in jCards.js (or the related expr function in gameUi.js) to exclude cards in the "SPECIAL" zone
  - updating clickCardFunction in gameUi.js to apply selectionFunction to a div and not a cardId
  - refactoring the decision code to use HTML elements instead of the "temp" cardDivId names; probably adding a map to crosswalk the cardIds to the HTML elements for things like the validCombinations code
- There is still some decision-specific code in gameUi.js, like decisionFunction and cleanupDecision. I may be removing some of this when I start to implement compound decisions.
- Because some of the arrays were replaced with maps in CardSelectionDecision (decisions.js, starting in line 110), I added an element called orderedCardDivIds. The server does not send cards to be selected in an intentional order, and I don't intend to start. This element exists only so that if you refresh the UI mid-decision, you won't get the cards in an entirely different order. For some decisions, the client may want to apply a sorting algorithm instead.
- There is a "SelectAndInsertAction" class on the server that really only exists as a way to convert a "select action" decision into a multiple choice string decision. This is used for cards like [Dedication to Duty](https://www.trekcc.org/1e/?cardID=3850) where it's not ideal to send the user back to the main UI for a standard action selection. Long-term, this should be communicated to the client using a decision context value instead.
- Removing jQuery. I made a token effort to do this with the card selection dialog buttons but couldn't quickly figure it out, so decided to leave it alone.

andrewd18 commented on gemp-module/gemp-stccg-client/src/main/web/js/gemp-022/decisions.js:
```diff
@@ -0,0 +1,564 @@
+                return "Select a mission to seed cards under or remove seed cards from";
+            case "SELECT_OPTIONAL_RESPONSE_ACTION":
+                return "Optional responses";
+            case "SELECT_PHASE_ACTION":
+                let phaseName = getFriendlyPhaseName(gameState.currentPhase);
+                return "Play " + phaseName + " action" + ((decision.min == 0) ? " or pass" : "");
```
You need to wrap the `let` and `play` lines here in `{}` brackets or it will crash when it gets to the `let`.

andrewd18 commented on gemp-module/gemp-stccg-client/src/main/web/js/gemp-022/decisions.js:
```diff
@@ -0,0 +1,564 @@
+        }
+        return false;
+    } else {
+        console.error("Unexpected elementType in getUseDialog: '" + elementType + "'");
```
`elementType` is not defined

andrewd18 commented on gemp-module/gemp-stccg-client/src/main/web/js/gemp-022/decisions.js:
```diff
@@ -0,0 +1,564 @@
+        }
+
+        // Reset focus
+        $(':button').blur();
```
Please use an `#id` lookup, class if you really have to

andrewd18 commented on gemp-module/gemp-stccg-client/src/main/web/js/gemp-022/decisions.js:
```diff
@@ -0,0 +1,564 @@
+        this.allowSelection();
+    }
+
+    respondToCardSelection(cardDivId, event) {
+        let cardDiv = getCardDivFromId(cardDivId);
```
event parameter is never used

cardDiv is never used

datanoh replied:
Removed both. the selectionFunction function that calls this still has the "event" parameter, which seems to be necessary for the action selections. I didn't really touch the "event" parameter in my edits except to remove it in response to this comment.

andrewd18 commented on gemp-module/gemp-stccg-client/src/main/web/js/gemp-022/decisions.js:
```diff
@@ -0,0 +1,564 @@
+    });
+
+    gameUi.smallDialog.dialog("open");
+    $('.ui-dialog :button').blur();
```
Use ID or class lookup please

datanoh replied:
Was part of the code before my edits.

andrewd18 commented on gemp-module/gemp-stccg-client/src/main/web/js/gemp-022/decisions.js:
```diff
@@ -0,0 +1,564 @@
+    }
+
+    gameUi.smallDialog.dialog("open");
+    $('.ui-dialog :button').blur();
```
Use ID or class lookup, please

andrewd18 commented on gemp-module/gemp-stccg-client/src/main/web/js/gemp-022/actionSelectionDecisions.js:
Please add a noop test case for `actionSelectionDecisions.js` in the tests folder, to ensure we at least check the functions for validity.  See `actionResults.test.js` for an example.

andrewd18 commented on gemp-module/gemp-stccg-client/src/main/web/js/gemp-022/decisions.js:
Please add a noop test case for `decisions.js` in the tests folder, to ensure we at least check the functions for validity.  See `actionResults.test.js` for an example.

datanoh commented on gemp-module/gemp-stccg-client/src/main/web/js/gemp-022/actionSelectionDecisions.js:
```diff
@@ -0,0 +1,342 @@
+        this.gameUi.hand.layoutCards();
+        this.allowSelection();
+        goDing(this.gameUi);
+        $(':button').blur();
```
Was part of the existing code before my edits.

datanoh commented:
@andrewd18 Thanks! Addressed your comments. I updated one of the .blur() lines because I realized I had accidentally changed the functionality in my PR. Otherwise, I'm not planning to update the .blur() lines or createActionChoiceContextMenu, since I didn't make meaningful edits to either in this PR.

https://github.com/trekcc-ops/gemp-stccg/blob/7b5307f9c954d407c48db944791ca9cf0e2c5ff5/gemp-module/gemp-stccg-client/src/main/web/js/gemp-022/gameUi.js#L1924-L1927

https://github.com/trekcc-ops/gemp-stccg/blob/7b5307f9c954d407c48db944791ca9cf0e2c5ff5/gemp-module/gemp-stccg-client/src/main/web/js/gemp-022/gameUi.js#L2088-L2134
